File: js/src/nft/retrieveNftOwner.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.retrieveNftOwner = void 0;
const web3_js_1 = require("@solana/web3.js");
const getDomainMint_1 = require("./getDomainMint");
const spl_token_1 = require("@solana/spl-token");
/**
 * This function can be used to retrieve the owner of a tokenized domain name
 *
 * @param connection The solana connection object to the RPC node
 * @param nameAccount The key of the domain name
 * @returns
 */
const retrieveNftOwner = (connection, nameAccount) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        const mint = (0, getDomainMint_1.getDomainMint)(nameAccount);
        const mintInfo = yield (0, spl_token_1.getMint)(connection, mint);
        if (mintInfo.supply.toString() === "0") {
            return undefined;
        }
        const filters = [
            {
                memcmp: {
                    offset: 0,
                    bytes: mint.toBase58(),
                },
            },
            {
                memcmp: {
                    offset: 64,
                    bytes: "2",
                },
            },
            { dataSize: 165 },
        ];
        const result = yield connection.getProgramAccounts(spl_token_1.TOKEN_PROGRAM_ID, {
            filters,
        });
        if (result.length != 1) {
            return undefined;
        }
        return new web3_js_1.PublicKey(result[0].account.data.slice(32, 64));
    }
    catch (_a) {
        return undefined;
    }
});
exports.retrieveNftOwner = retrieveNftOwner;

File: python/src/nft/retrieve_nft_owner.py
from solana.rpc.async_api import AsyncClient
from solders.pubkey import Pubkey
from solders.rpc.errors import InvalidParamsMessage

from exception import SolanaJSONRPCException
from nft.AccountLayout import ACCOUNT_LAYOUT
from nft.get_domain_mint import get_domain_mint


async def retrieve_nft_owner(
    connection: AsyncClient, name_account: Pubkey
) -> Pubkey | None:
    """
    Retrieve the owner of an NFT.
    :param connection:
    :param name_account: the account of the NFT
    :return: human-readable name of the owner
    """
    try:
        mint = get_domain_mint(name_account)

        largest_accounts = await connection.get_token_largest_accounts(mint)
        if isinstance(largest_accounts, InvalidParamsMessage):  # type: ignore[unreachable]
            return None  # type: ignore[unreachable]
        if not largest_accounts or len(largest_accounts.value) == 0:
            return None

        largest_account_info = await connection.get_account_info(
            largest_accounts.value[0].address
        )
        if not largest_account_info or not largest_account_info.value:
            return None

        decoded = ACCOUNT_LAYOUT.parse(largest_account_info.value.data)
        if decoded.amount == 1:
            return Pubkey(decoded.owner)
        return None
    except SolanaJSONRPCException as err:
        if err.code == -32602:
            # Mint does not exist
            return None
        raise err
    except Exception as err:
        raise err
